Add new useVirtualFocus hook, for use in Combobox

it-vegard opened the pull request:
Might be useful for other components later, but will scope it for Combobox at first.

### Description

The useVirtualFocus hook takes in a ref pointing to the container of the elements to use virtual focus on.
E.g. the `ul` element in the dropdown list in Combobox.

The hook will maintain the reference to the currently "focused" child element, and provides methods for moving focus up and down by one, jump to the last element or to the top (currently setting the index to -1 to signal virtual focus being removed) or setting virtual focus to a specific element.

The hook also returns the element currently receiving virtual focus. This can then be used to get the ID, for use in e.g. `aria-activeDescendant`.

### Change summary

Added useVirtualHook

### PR Checklist TODOs 📝 

#### New component? ✨

- Add JSDoc to component (see other components for reference)

#### Versioning 🏷️

- Run `yarn changeset` to generate a version-entry for change.

it-vegard committed:
Add new useVirtualFocus hook, for use in Combobox. Might be useful for other components later, but will scope it for Combobox at first.

it-vegard committed:
Create filteredOptionsUtils, which has utiltity functions for creating IDs and for text lookups in lists.

it-vegard committed:
Get rid of data-value, and use filteredOptionsMap to look up the value of the currently (virtually) focused element instead

This allows use of e.g. numbers as options, which data-value didn't.

it-vegard committed:
Using useState instead of useRef to handle filteredOptionsRef

it-vegard committed:
Merge branch 'main' into feature/useVirtualFocus-hook-for-combobox

it-vegard committed:
Added changeset

it-vegard committed:
Merge branch 'main' into feature/useVirtualFocus-hook-for-combobox

it-vegard committed:
Merge branch 'main' into feature/useVirtualFocus-hook-for-combobox

# Conflicts:
#	@navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx
#	@navikt/core/react/src/form/combobox/Input/Input.tsx
#	yarn.lock

it-vegard committed:
Force option id to be lower case, to avoid different casing from being interpreted as being different from the existing options

it-vegard committed:
Merge branch 'feature/useVirtualFocus-hook-for-combobox' of https://github.com/navikt/aksel into feature/useVirtualFocus-hook-for-combobox

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filtered-options-util.ts:
```diff
@@ -0,0 +1,38 @@
+const normalizeText = (text: string): string =>
+  typeof text === "string" ? `${text}`.toLowerCase().trim() : "";
```
Trengs template literal?

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -151,102 +152,38 @@ export const FilteredOptionsProvider = ({ children, value: props }) => {
-    filteredOptionsRef,
-    filteredOptionsIndex,
-    setFilteredOptionsIndex,
+    filteredOptionsRef: setFilteredOptionsRef,
```
Hvorfor ikke bruke samme navn?

it-vegard replied:
Dette skyldes endring fra å bruke useRef til useState for å holde på filteredOptionsRef, som foreslått av @KenAJoh. Husker ikke helt argumentasjonen for det, men mener det gikk på at vi ønsket at endringer i elementet skulle føre til at useVirtualFocus re-intialiserte seg (tror jeg?).

Da sitter vi med både filteredOptionsRef og setFilteredOptionsRef, men utenfor denne filen vil jeg gjerne fremdeles se på det som en ref, og dermed renamet jeg ikke den.

Men er åpen for å bli overbevist om det motsatte.

HalvorHaugan replied:
Tror ikke ref-en blir satt på nytt ved endringer i elementet, bare når det mountes og unmountes.

it-vegard replied:
Jeg kan nok ha misforstått noe her. Jeg skal sjekke med Ken hva som var greien her.

KenAJoh replied:
Vil være risiko for at `useVirtualFocus` ikke blir oppdatert riktig hvis `filteredOptionsRef` er en `React.useRef` da den ikke fører til state-updates ved endring. 
```
const virtualFocus = useVirtualFocus(filteredOptionsRef);
```
`filteredOptionsRef` vil her potensielt være `null` ved første kjøring, men vil kjøres på nytt når `setFilteredOptionsRef(ref)` blir satt.

I praksis blir sikkert `useVirtualFocus` kjørt på nytt uansett da andre state-updates oppdaterer `useVirtualFocus`, så er bare initial mount som blir påvirket. 

(Har ikke testet om forskjellen påvirker faktisk kode da)

HalvorHaugan replied:
Kan kjøpe den, spesielt siden useVirtualFocus ikke ligger i en useEffect.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -81,19 +68,32 @@ export const FilteredOptionsProvider = ({ children, value: props }) => {
+  const filteredOptionsMap = useMemo(
+    () =>
+      options.reduce(
+        (map: Map<string, string>, _option: string) => ({
```
Feil type på `map`. Hvis du definerer typer på `FilteredOptionsProvider` sine props, så slipper du å definere noen typer her.

it-vegard replied:
Helt korrekt. 👍  Definerte typer for props, og gjenbrukte fra ComboboxProps, så oppdaget jeg litt feiltyping rundt om.

it-vegard committed:
Fix bug where it was not possible to select multiple custom (new) options in a multiselect

selectedOptionsContext is not available inside customOptionsContext, so need to take isMultiSelect from props instead

it-vegard committed:
Switch to use toLocalLowerCase and remove uneccessary use of string literal

it-vegard committed:
type mismatch

it-vegard committed:
Merge branch 'main' into feature/useVirtualFocus-hook-for-combobox

# Conflicts:
#	yarn.lock

it-vegard committed:
Clean up types - reuse from ComboboxProps type definition, and fix resulting type mismatches

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -151,102 +170,38 @@ export const FilteredOptionsProvider = ({ children, value: props }) => {
-    toggleIsListOpen(true);
-    setFilteredOptionsIndex(lastIndex);
-  }, [filteredOptions.length, toggleIsListOpen]);
+  // TODO: Re-write or remove after re-write?
```
Skal disse TODO-kommentarene bli værende i koden, eller har du glemt å fjerne/gjøre dem?

it-vegard replied:
Akkurat denne tror jeg ikke vi kom utenom likevel, så her skal jeg fjerne TODO-en. De to neste kan for så vidt fremdeles gjøres. Skal se på det.

it-vegard committed:
Rename filteredOptionsRef variable returned from filteredOptionsContext to reflect that it is now actually a setState-function

it-vegard committed:
Expose virtualFocus as an object through filteredOptionsContext, instead of each separate method.

Done to remove need to wrap moveFocusUp and moveFocusDown in order to toggle isListOpen

it-vegard committed:
Merge branch 'main' into feature/useVirtualFocus-hook-for-combobox

larseirikhansen commented on @navikt/core/react/src/form/combobox/FilteredOptions/filtered-options-util.ts:
Liker denne filen, smart å trekke ut utils

larseirikhansen reviewed (approved):
Synes endringene ser ryddige ut, lets 🚢 🇮🇹 Så vi får gjort ferdig neste PR 👍
